src/basic/alloc-util.h

```diff
@@ -170,3 +171,50 @@ void* greedy_realloc0(void **p, size_t *allocated, size_t need, size_t size);
 #else
 #  define msan_unpoison(r, s)
 #endif
+
+/* The following implements a logic for automatically erasing a specific memory area when the current scope
+ * is left. The AUTO_ERASE_MEMORY() macro takes a pointer and a size, of which it takes references. When the
+ * pointer is non-NULL when the current scope is left the pointer it points to is automatically erased and
+ * freed. Usecase:
+ *
+ *
+ * {
+ *         …
+ *         void *secret_key;
+ *         size_t secret_key_size;
+ *         …
+ *         secret_key_size = 256;
+ *         secret_key = malloc(secret_key_size);
+ *         if (!secret_key)
+ *                 return log_oom();
+ *
+ *         AUTO_ERASE_MEMORY(secret_key, secret_key_size);
+ *         …
+ *         genuine_random_bytes(secret_key, secret_key_size);
+ *         …
+ * } // memory of secret_key will be erased and freed here, automatically.
+ *
+ */
+
+struct auto_free_erase_info {
+        void **ptr;
+        size_t *size;
+};
```

Comment:
I think this is a good cleanup, but the I don't like the macro definition. We have too many macros anyway. In particular, the annotation about freeing is split from the definition of the variable. I think this will lead to leaks. Why not just define the structure explicitly?
```C
typedef memory_alloc {
   union { char *charp; void *voidp; ... };
   size_t size;
};
static void free_erase_memory_alloc(memory_alloc *a) {
       if (a->voidp) {
                ...
        }
}

_cleanup_(free_erase_memory_alloc) memory_alloc foo = {};

foo->size = 128;
foo->voidp = malloc(128);
strcmp(foo->charp, "xxx", 128);
```

Reply:
for me making this typesafe was key really, which is the hard part afaics... i.e. we usually allocate some structure or so, and i don't see how we could do this without macro magic?

Reply:
By making a strong assumption on use of glibc and that its [internals for memory allocation](https://sourceware.org/glibc/wiki/MallocInternals) stay the same, the allocation size needed for clearing could be read directly using just the pointer returned by `malloc()`. Then the ordinary gcc cleanup extension could be used without macros.

We could also request for an extension to `free()` which also clears the memory, like `cfree()`. I think this would be useful for lots of other projects with similar needs.

Reply:
@poettering I opened a new [bug 24775](https://sourceware.org/bugzilla/show_bug.cgi?id=24775) for glibc requesting `clear_and_free()`.

Reply:
@poettering Glibc devs would rather have MSan fixed so I closed the bug for glibc. I briefly looked at MSan source and it seems to intercept `malloc_usable_size()` so that it returns the original allocated amount. Thus MSan should not complain if we zero the amount returned by `malloc_usable_size()`. Do you have a test case for this? @keszybz 's #12965 built with sanitizers successfully.

Reply:
>  @keszybz 's #12965 built with sanitizers successfully.

It didn't. On Azure Pipelines (which is the only place where the code is built and run with MSan) [It failed to compile with](https://dev.azure.com/evvers/systemd-systemd/_build/results?buildId=1274)
```
In file included from ../src/basic/socket-util.c:25:
../src/basic/memory-util.h:94:9: error: implicit declaration of function 'explicit_bzero' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        explicit_bzero(p, sizeof(char));
        ^
../src/basic/memory-util.h:94:9: warning: this function declaration is not a prototype [-Wstrict-prototypes]
1 warning and 1 error generated.
FAILED: clang -Isrc/basic/a6ba3eb@@basic@sta -Isrc/basic -I../src/basic -Isrc/shared -I../src/shared -Isrc/systemd -I../src/systemd -Isrc/journal -I../src/journal -Isrc/journal-remote -I../src/journal-remote -Isrc/nspawn -I../src/nspawn -Isrc/resolve -I../src/resolve -Isrc/timesync -I../src/timesync -I../src/time-wait-sync -Isrc/login -I../src/login -Isrc/udev -I../src/udev -Isrc/libudev -I../src/libudev -Isrc/core -I../src/core -Isrc/shutdown -I../src/shutdown -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -Isrc/libsystemd-network -I../src/libsystemd-network -I. -I../ -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -g -fsanitize=fuzzer-no-link -Wextra -Werror=undef -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wfloat-equal -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Werror=shift-count-overflow -Wdate-time -Wnested-externs -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result '-Wno-error=#warnings' -Wno-string-plus-int -Wno-error=nonnull -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -Wno-typedef-redefinition -Wno-gnu-variable-sized-type-not-at-end -Werror=shadow -include config.h -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fPIC -pthread -fvisibility=default -MD -MQ 'src/basic/a6ba3eb@@basic@sta/alloc-util.c.o' -MF 'src/basic/a6ba3eb@@basic@sta/alloc-util.c.o.d' -o 'src/basic/a6ba3eb@@basic@sta/alloc-util.c.o' -c ../src/basic/alloc-util.c
In file included from ../src/basic/alloc-util.c:9:
../src/basic/memory-util.h:94:9: error: implicit declaration of function 'explicit_bzero' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        explicit_bzero(p, sizeof(char));
        ^
../src/basic/memory-util.h:94:9: warning: this function declaration is not a prototype [-Wstrict-prototypes]
1 warning and 1 error generated.
```
which also means that the fuzzers are broken, which in turn prevented fuzzit from running them against the corpora under ASan and UBsan.